### PR TITLE
Interactivity reorganization and improvement. 

### PR DESCRIPTION
**Pull request by Giggitybyte**
# Summary
The primary goal of this PR is to clean up and improve extension methods, enums, and return types from `DSharpPlus.Interactivity`.

# Details
In addition to some other minor clean up, all extension methods and enums have been gathered into one place to allow for easier maintenance and readability. A few quality of life improvements have been made as well.

# Changes proposed
* Improve extension methods.
  - All extension methods are now under the new *Extensions* namespace.
  - All extensions with a single awaitable statement now pass their task through instead of `await`ing.
  - Improved XML docs for all extension methods.
  - Improved variable naming for all extension methods.
  - Added missing null checks.
  - Config parameter for `UseInteractivity` is now optional.
  - Config parameter for `UseInteractivityAsync` is now optional.
  - The aforementioned extensions will now create a default `InteractivityConfiguration` instance if none is provided.

* Improve enums.
  - Moved `PaginationDeletion` from *PaginationBehaviour.cs* into its own file.
  - Moved `SplitType` from *InteractivityExtension.cs* into its own file under the *Enum* namespace.
  - Removed unnecessary `Flags` attribute on `PollBehaviour`.
  - Improved XML documentation for all enums.

* Remove *Concurrency* namespace.
  - Deleted `ConcurrentHashSet`
  - Deleted `PlatformHelper`
  - Replaced the above with a NuGet package which contains the *exact* same types from the original author.

* Replace all array types with an `IEnumerable<>` equivalent.
  - `pages` parameter of `SendPaginatedMessageAsync` is now of type `IEnumerable<Page>` instead of `Page[]`.
  - `emojis` parameter of `DoPollAsync` is now of type `IEnumerable<DiscordEmoji>` instead of `DiscordEmoji[]`.
  - `GeneratePagesInString` now returns `IEnumerable<Page>` instead of `Page[]`.
  - `GeneratePagesInEmbed` now returns `IEnumerable<Page>` instead of `Page[]`.

Hopefully I haven't missed anything.

# Notes
There's more that could be improved, but I wanted to avoid any *major* breaking changes with this PR.
Perhaps after 4.0 releases, I'll PR those additional changes.

### CODE REVIEW

File: DSharpPlus.Interactivity/Extensions/ClientExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        /// <summary>
+        /// Enables interactivity for this <see cref="DiscordClient"/> instance.
+        /// </summary>
+        /// <param name="client">The client to enable interactivity for.</param>
+        /// <param name="configuration">A configuration instance. Default configuration values will be used if none is provided.</param>
+        /// <returns>A brand new <see cref="InteractivityExtension"/> instance.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if interactivity has already been enabled for the client instance.</exception>
+        public static InteractivityExtension UseInteractivity(this DiscordClient client, InteractivityConfiguration configuration = null)
+        {
+            if (client.GetExtension<InteractivityExtension>() != null) throw new InvalidOperationException("Interactivity is already enabled for this client.");
```
Comment (Neuheit) [inline DSharpPlus.Interactivity/Extensions/ClientExtensions.cs:23]:
```suggestion
            if (client.GetExtension<InteractivityExtension>() != null) throw new InvalidOperationException("Interactivity is already enabled for this {(client._isShard ? "shard" : "client")}.");
```

Comment (Neuheit) [reply DSharpPlus.Interactivity/Extensions/ClientExtensions.cs:23]: To keep things consistent.

File: DSharpPlus.Interactivity/Extensions/ClientExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        /// <summary>
+        /// Retrieves a <see cref="InteractivityExtension"/> instance for each shard.
+        /// </summary>
+        /// <param name="client">The shard client to retrieve interactivity instances from.</param>
+        /// <returns>A dictionary containing <see cref="InteractivityExtension"/> instances for each shard.</returns>
+        public static IReadOnlyDictionary<int, InteractivityExtension> GetInteractivity(this DiscordShardedClient client)
+        {
+            var extensions = new Dictionary<int, InteractivityExtension>();
+            client.InitializeShardsAsync().ConfigureAwait(false).GetAwaiter().GetResult();
```
Comment (Neuheit) [inline DSharpPlus.Interactivity/Extensions/ClientExtensions.cs:68]: This reminds me from another PR, we should consider making this awaited and have the method be async, along with the same methods in the other extensions. cc @Emzi0767

File: DSharpPlus.Interactivity/Extensions/ClientExtensions.cs (hunk 1 above)
Comment (Giggitybyte) [reply DSharpPlus.Interactivity/Extensions/ClientExtensions.cs:23]: Whoops, I intended to go back and do that. Thanks for catching this!

**Comment by Neuheit** [comment]
Whoops that was my bad haha.

File: DSharpPlus.Interactivity/Extensions/ClientExtensions.cs (hunk 2 above)
Comment (Emzi0767) [reply DSharpPlus.Interactivity/Extensions/ClientExtensions.cs:68]: Yeah I agree. If we're making it async, we should make all the other methods async as well. While normally out-of-scope, I'll allow that for this PR.

Comment (Giggitybyte) [reply DSharpPlus.Interactivity/Extensions/ClientExtensions.cs]: `DSharpPlus.Lavalink` and *VoiceNext* do not have an extension for `DiscordShardedClient`. However, both *CommandsNext* and *Interactivity* do. 
I've made the needed changes to the interactivity extension method in this PR, but I'd prefer to bring the other three DSharpPlus extensions in line with a dedicated PR.

**Review by Neuheit — APPROVED** [review]

**Pull request merged**